corehq/apps/app_manager/fields.py

```diff
@@ -116,3 +139,373 @@ def get_app_sources(domain):
         }
         for app in apps
     }
+
+
+class ApplicationDataRMIHelper(object):
+    """ApplicationDataRMIHelper is meant to generate the response for the
+    djangoRMI methods required to initialize form controlled by
+    hq.app_data_drilldown.ng.js.
+
+    Note / todo: This Helper should be merged with ApplicationDataSourceUIHelper.
+    Holding off to a different PR, as I want to isolate QA to just the exports
+    (the first thing to use this) --Biyeun
+
+    """
+    UNKNOWN_SOURCE = '_unknown'
+
+    APP_TYPE_ALL = 'all'
+    APP_TYPE_DELETED = 'deleted'
+    APP_TYPE_REMOTE = 'remote'
+    APP_TYPE_NONE = 'no_app'
+
+    def __init__(self, domain, as_dict=True, form_placeholders=None, case_placeholders=None):
+        self.domain = domain
+        self.as_dict = as_dict
+        default_form_placeholder = AppFormRMIPlaceholder(
+            application=_("Select Application"),
+            module=_("Select Module"),
+            form=_("Select Form"),
+        )
+        self.form_placeholders = form_placeholders or default_form_placeholder
+        default_case_placeholder = AppCaseRMIPlaceholder(
+            application=_("Select Application"),
+            case_type=_("Select Case Type"),
+        )
+        self.case_placeholders = case_placeholders or default_case_placeholder
+        if self.as_dict:
+            self.form_placeholders = self.form_placeholders._asdict()
+            self.case_placeholders = self.case_placeholders._asdict()
+
+    def _get_unknown_form_possibilities(self):
+        possibilities = collections.defaultdict(list)
+        for app in get_exports_by_application(self.domain):
+            # index by xmlns
+            x = app['value']
+            x['has_app'] = True
+            possibilities[app['key'][2]].append(x)
+        return possibilities
+
+    def _attach_unknown_suggestions(self, unknown_forms):
+        """If there are any unknown forms, try and find the best possible matches
+        from deleted apps or copied apps. If no suggestion is found, say so
+        but provide the xmlns.
+        """
+        if unknown_forms:
+            possibilities = self._get_unknown_form_possibilities()
+
+            class AppCache(dict):
+                def __init__(self, domain):
+                    super(AppCache, self).__init__()
+                    self.domain = domain
+
+                def __getitem__(self, item):
+                    if item not in self:
+                        try:
+                            self[item] = get_app(app_id=item, domain=self.domain)
+                        except Http404:
+                            pass
+                    return super(AppCache, self).__getitem__(item)
+
+            app_cache = AppCache(self.domain)
+
+            for form in unknown_forms:
```

Comment:
this code would be much easier to parse if you took [this suggestion](https://github.com/dimagi/commcare-hq/pull/8399/files#r45446365) and made these objects real classes with a clear api